MCUboot DTS integration

carlescufi opened the pull request.

carlescufi committed:
dts: Move dts.cmake to cmake/

Move the DTS CMake file to the cmake/ folder, where all other .cmake
files live.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
dts: mcuboot: Support for common overlays and mcuboot

When the Kconfig BOOTLOADER_MCUBOOT is selected, an overlay to place the
image at the slot0 location is required. In order to avoid having to do
this manually for all samples when targetting MCUboot, include the logic
inside the dts.cmake script to prepend a new common.dts file that then
conditionally includes mcuboot.overlay.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

nvlsianpu commented:
looks good

nashif reviewed (changes requested):
can we have some sample that uses that? we want to see this option CONFIG_BOOTLOADER_MCUBOOT being used somewhere to make sure we are not missing any build issues.

carlescufi commented:
@nashif will do!

nashif commented:
agree

mbolivar commented on misc/Kconfig:
```diff
@@ -317,8 +317,15 @@ config BOOTLOADER_MCUBOOT
+	  - Activating SW_VECTOR_RELAY on Cortex-M0 targets with no built-in
+	  vector relocation mechanisms
+	  - Including subsys/dfu/boot/mcuboot.overlay when building the Device
+	  Tree in order to place and link the image at the slot0 offset
```
PR overall looks good, just some nits:

- The alignment in these help bullet list items looks off
- Do we really want to start putting DTS files under subsys/dfu/, instead of under dts/?

I know we have per-board dts files in the board directories, but that is because we want those to be able to be self-contained eventually. There is already a dts/common/ directory, why not use that? (Honest question, I'm open to either solution).

carlescufi replied:
>The alignment in these help bullet list items looks off

See comment below.

>Do we really want to start putting DTS files under subsys/dfu/, instead of under dts/?

I really don't mind, I like your suggestion so I'll do that.

carlescufi replied:
Actually on second thought, I think the reason I placed it outside of `dts/` is because in there there are only `.dtsi` files and not `.dts`.

mbolivar replied:
^^

@galak do you care if dts overlay files go into dts/common alongside the dtsi files? Or is that directory only for dtsi?

dbkinder commented on doc/application/application.rst:
```diff
@@ -928,6 +928,10 @@ checking the following until one is found, in order:
 
+Additionally, the build system will automatically include the required overlay
+for MCUboot chain-loading (located in :file:`subsys/dfu/boot/mcuboot.overlay`)
+whenever the :makevar:`CONFIG_BOOTLOADER_MCUBOOT` is set using Kconfig.
```
I suspect this should be ``:option:`CONFIG_BOOTLOADER_MCUBOOT` ``

dbkinder commented on misc/Kconfig:
```diff
@@ -317,8 +317,15 @@ config BOOTLOADER_MCUBOOT
+	  This sets several required build system and Device Tree options in
+	  order for the image generated to be bootable using the MCUboot open
+	  source bootloader. Currently this includes:
+	  - Setting TEXT_LOAD_OFFSET to a default value that allows space for
```
need a blank line before a list and the continuation lines need to be indented:

    source bootloader. Currently this includes:

    - Setting TEXT_LOAD_OFFSET to a default value that allows space for
      the MCUboot image header
    - Activating SW_VECTOR_RELAY on Cortex-M0 targets with no built-in
      vector relocation mechanisms
    - Including subsys/dfu/boot/mcuboot.overlay when building the Device
      Tree in order to place and link the image at the slot0 offset

carlescufi replied:
I did that but then the build fails, seems like Kconfig option help text requires consistent alignment

nashif replied:
you need to shift the bullets to the right I guess, look at this as if it was rst

dbkinder reviewed (changes requested):
a few formatting tweaks needed

dbkinder reviewed (approved):
yea

carlescufi committed:
tests: build: Test build with MCUboot support

Add a new build variant that builds for Nordic nRF52832 with the new
MCUboot CONFIG_BOOTLOADER_MCUBOOT Kconfig variable set.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@nashif test build added

SebastianBoe commented on doc/application/application.rst:
```diff
@@ -928,6 +928,10 @@ checking the following until one is found, in order:
+Additionally, the build system will automatically include the required overlay
+for MCUboot chain-loading (located in :file:`dts/comon/mcuboot.overlay`)
+whenever the :option:`CONFIG_BOOTLOADER_MCUBOOT` is set using Kconfig.
+
```
An alternative solution would be to have 1 DTS source file for all Zephyr builds that included other DTS sources, just like we do with the 1 linker script that is shared between all ARM builds.

Something like this:

```
#include <device.dts>

#ifdef CONFIG_BOOTLOADER_MCUBOOT
#include <dts/comon/mcuboot.overlay>
#endif

```

Then we could get rid of the pollution to the core build system.

carlescufi replied:
I agree with this, will implement now

MaureenHelm commented on tests/misc/test_build/testcase.yaml:
```diff
@@ -16,3 +16,8 @@ tests:
       - CONFIG_GDB_SERVER=y
+  test_mcuboot_bootloader:
+    build_only: true
+    platform_whitelist: nrf52_pca10040
```
Can we also include a platform that uses SW_VECTOR_RELAY?

carlescufi replied:
Will do

carlescufi commented on cmake/dts.cmake:
```diff
@@ -21,6 +22,11 @@ if(CONFIG_HAS_DTS)
   endif()
 
+  # Include MCUboot overlay file if set in Kconfig
+  if(CONFIG_BOOTLOADER_MCUBOOT)
```
I have checked against other similar variables and I must disagree in this case. I think this is aligned not only against the rest of bootloader options, but also against other similar features that depend on external components.

carlescufi commented:
@SebastianBoe, @MaureenHelm comments addressed

SebastianBoe reviewed (approved):
Much better.
